juju/juju
# Allow NotifyTarget to return an error

The notify target in the raft lease store was forcing errors to be
handled at the lowest point of the implementation. This meant that there
is no way to handle this in a different manner. Additionally, it makes
testing very difficult as you now have to test that the logger has
changed to ensure that something happened.

The code now returns an error to improve the tests and
we log at a much higher level. One that can decide how best to handle
the error.

The implementation of notify target now makes a lot more sense in terms of the
visitor pattern, which is what this was trying to replicate.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju model-config -m controller logging-config="<root>=INFO;juju.worker.lease.raft=TRACE;juju.core.raftlease=TRACE;juju.worker.globalclockupdater.raft=TRACE;juju.worker.raft=TRACE"
$ juju enable-ha
$ juju deploy ubuntu
$ juju debug-log -m controller 
```

Ensure we don't get any errors.
